ansible/config_sonic_basedon_testbed.yml

```diff
@@ -414,15 +414,30 @@
     when: proxy_env is defined and deploy is defined and deploy|bool == true
 
   - block:
+      - name: Load variables from topology file for topo_{{ topo }}.yml
+        include_vars: "vars/topo_{{ topo }}.yml"
```

Comment:
The `topo_facts` module in line 88 is able to load the topology details. This include step is unnecessary and potentially harmful.
If simply include the topology definition as a variable file, the ansible variable space could be polluted and could cause variable name conflicts which are hard to debug.

Reply:
line 88 doesn't load topology yml right?

  - topo_facts: topo={{ topo }} hwsku={{ hwsku }} testbed_name={{ testbed_name | default(None) }} asics_present={{ asics_present | default([]) }} card_type={{ card_type | default('fixed') }}
    delegate_to: localhost

Should I move this loading to the beginning of this file to reduce the variable space polluted concerns?

Reply:
Probably the better idea is to remove this `include_vars`. For later code needs information in the topo definition yaml, refer to them in results returned by the `topo_facts` module. The `topo_facts` module is exactly for this purpose.

Reply:
How to include this information in topo_facts? Would topo file get read before topo_facts?

Reply:
config_sonic_basedon_testbed.yml serves as an entry points in testbed-cli.sh

ansible/testbed-cli.sh:  ansible-playbook -i "$inventory" config_sonic_basedon_testbed.yml --vault-password-file="$passfile" -l "$duts" -e testbed_name="$testbed_name" -e testbed_file=$tbfile -e vm_file=$vmfile -e local_minigraph=true $@
ansible/testbed-cli.sh:  ansible-playbook -i "$inventory" config_sonic_basedon_testbed.yml --vault-password-file="$passfile" -l "$duts" -e testbed_name="$testbed_name" -e testbed_file=$tbfile -e vm_file=$vmfile -e deploy=true -e save=true $@
ansible/testbed-cli.sh:  ansible-playbook -i "$inventory" --diff --connection=local --check config_sonic_basedon_testbed.yml --vault-password-file="$passfile" -l "$duts" -e testbed_name="$testbed_name" -e testbed_file=$tbfile -e vm_file=$vmfile -e local_minigraph=true $@

We have yml file to load_var from topology file as the following which would be triggered via 

https://github.com/sonic-net/sonic-mgmt/blob/master/ansible/roles/vm_set/tasks/start.yml#L6

which would be triggered by config_sonic_basedon_testbed.yml, a.k.a later down the road, this topology file would be loaded.

Therefore, I don't feel the variable pollution would be a concern for loading topology files in this yml file.

Reply:
The `topo_facts` module basically parses topology and store all topology information in a few dicts which can be accessed from ansible explicitly. Explicit is better than implicit.
https://github.com/sonic-net/sonic-mgmt/blob/master/ansible/library/topo_facts.py#L392

But anyway, I noticed that this inclue_vars method was widely used in this repo. It's OK to use it here.